source/common/upstream/load_balancer_impl.cc

```diff
@@ -14,18 +14,49 @@ const std::vector<HostPtr>& LoadBalancerBase::hostsToUse() {
     return host_set_.hosts();
   }
 
-  uint64_t threshold =
+  uint64_t global_panic_threshold =
       std::min(100UL, runtime_.snapshot().getInteger("upstream.healthy_panic_threshold", 50));
   double healthy_percent =
       (host_set_.healthyHosts().size() / static_cast<double>(host_set_.hosts().size())) * 100;
 
   // If the % of healthy hosts in the cluster is less than our panic threshold, we use all hosts.
-  if (healthy_percent < threshold) {
+  if (healthy_percent < global_panic_threshold) {
     stats_.upstream_rq_lb_healthy_panic_.inc();
     return host_set_.hosts();
-  } else {
-    return host_set_.healthyHosts();
   }
+
+  // Check if we need to perform zone aware routing, by default disabled.
+  if (runtime_.snapshot().featureEnabled("upstream.zone_routing.enabled", 0)) {
+    double zone_to_all_percent =
+        100.0 * host_set_.localZoneHealthyHosts().size() / host_set_.healthyHosts().size();
+    double expected_percent = 100.0 / stats_.upstream_zone_count_.value();
+
+    uint64_t zone_percent_diff =
+        runtime_.snapshot().getInteger("upstream.zone_routing.percent_diff", 0);
+
+    // Hosts should be roughly equally distributed between zones.
+    if (std::abs(zone_to_all_percent - expected_percent) > zone_percent_diff) {
```

Comment:
Agree recalculating is not optimal, but if we do not do recalculation then runtime change will be in effect only after membership changes.
I think calculations are not that intensive, for now extracting logic into method in HostSet::localZoneHealthyBalanced(), then we can easily make that lazy calculated if needed.